AMP-68857 - add JS device ID section (#789)
Amplitude-Developer-Docs/amplitude-dev-center: docs/data/sdks/javascript/index.md

amplitude.getInstance().init('API_KEY', 'USER_ID', { onExitPage: exitCallback });
```

### Device ID lifecycle

By default Amplitude Javascript SDK uses a base64 ID as device ID in order to have more compact string IDs. A 22-character base64 ID has the same range as a 128-bit UUID, which is represented as a 36 character string.

The SDK initializes the device ID in the following order, with the device ID being set to the first valid value encountered:

1. Device id in configuration on initialization
2. "amp_device_id" value from URL param if `configuration.deviceIdFromUrlParam` is true. Refer to [cross domain tracking](.#cross-domain-tracking-javascript) for more details
3. Device id in cookie storage. Refer to [cookie management](./#cookie-management) for more details
4. A randomly generated base64 ID

#### When does a device ID change?

A device ID changes in many scenarios:

- `setDeviceId()` is called explicitly.
- By default the SDK stores device IDs in cookies, so a device ID will change if a user clears cookies, use another device, or use privacy mode.

> - On initilization, a device ID is passed in from URL param `amp_device` when `deviceIdFromUrlParam` is enabled.

Reviewer: `amp_device_id`

Author: fixed

!!!Note "Amplitude Analytics SDKs share an identity store with Experiment SDKs"
    `setDeviceId` also updates the identity store to propagate new user info to experiment SDK and trigger a fetch if device ID has changed.

#### Set device ID

> By default, device IDs are randomly generated base64 IDs. You can define a custom device ID by setting it as a configuration option or by calling `setDeviceId`.

Reviewer: ```suggestion
By default, the device ID is randomly generated base64 ID. You can define a custom device ID by setting it as a configuration option or by calling `setDeviceId`.
```

```js
amplitude.getInstance().setDeviceId('DEVICE_ID');
```

!!!note
    Amplitude doesn't recommend defining your own device IDs unless you have your own system for tracking user devices. Make sure the `deviceId` you set is unique to prevent conflicts with other devices in your Amplitude data.
    It's best practice to use something like a UUID.

    [See an example](https://github.com/amplitude/Amplitude-Javascript/blob/master/src/uuid.js) of how to generate UUIDs with JavaScript.

> #### Get device ID

Reviewer 2: Was abbreviations.md removed intentionally?

Author: Nice catch. I deleted it accidentally.

> You can retrieve the device ID that Amplitude uses with `Amplitude.getInstance().getDeviceId()` This method can return `null` if a `deviceId` hasn't been generated yet.

Reviewer: ```suggestion
You can retrieve the device ID that Amplitude uses with `Amplitude.getInstance().getDeviceId()` or `Amplitude.getInstance('YOUR-INSTANCE-NAME').getDeviceId()` if you defined a custom instance name. This method can return `null` if a `deviceId` hasn't been generated yet.
```

#### Share current device ID to another instance

> Sometimes you have more than one Amplitude Javascript SDK instance setup and want to share the device ID across instances. To do so, here are two ways:

Reviewer: three ways?

Author: Delete this line "To do so, here are two ways:" to be more general so that it can be put into a include file.

amplitude.getInstance().init('API_KEY', 'USER_ID', { onExitPage: exitCallback });
```

### Device ID lifecycle

By default Amplitude Javascript SDK uses a base64id as device id in order to have more compact string IDs. A 22-character base64id has the same range as a 128-bit UUID, which is represented as a 36 character string.

> On instance initialization, the SDK initializes the device ID by the following order:

Reviewer: I feel we need to emphasize when it gets a value, it returns and ignores the others. Also at the end, we should add 4. base64id()

How about this: 
```During instance initialization, the SDK initializes the device ID based on a predefined order. The deviceId will be chosen when it encounters a value, prioritizing the first valid value it finds in the order.```

Author: Thanks for your review Alyssa. It is more clear to emphasize that the device id will be the first valid value. 

Fix.

1. Device id in configuration on initialization

> 2. Device id from url param if `configuration.deviceIdFromUrlParam` is true

Reviewer: 2. `amp_device_id` value from URL param if `configuration.deviceIdFromUrlParam` is true 
Also, link the cross-domain tracking at the end. 
Then we can remove the note.

Author: Fix as suggested

3. Device id in cookie storage

!!!Note
    If `deviceIdFromUrlParam` is `true`, then the SDK will parse the device ID value from the URL parameter "amp_device_id" if available. For example, device ID “123456789” is passed in from http://example.com/?amp_device_id=123456789

> The SDK stores device ID in cookie by default. The name of the cookie is composed of a "amp_" prefix and the first six digits of the API key. So if a user clears cookies, uses another device or uses browser privacy mode, the device ID will be different.

Reviewer: I feel we should but this into a Note.

Author: Delete this paragraph and add a link to cookie management at the end of #3:
3. Device id in cookie storage. Refer to [cookies management](./#cookie-management) for more details

![Javascript SDK cookie storage](../../../assets/images/sdk/js-cookies.png)

#### When does a device ID change?

A device ID changes in many scenarios:

- `setDeviceId()` is called explicitly.
- By default the SDK stores device IDs in cookies, so a device ID will change if a user clears cookies, use another device, or use privacy mode.
- On initilization, a device ID is passed in from URL param `amp_device` when `deviceIdFromUrlParam` is enabled.

!!!Note "Amplitude Analytics SDKs share an identity store with Experiment SDKs"
    `setDeviceId` also updates the identity store to propagate new user info to experiment SDK and trigger a fetch if device ID has changed.

#### Set device ID

> By default, device IDs are randomly generated base64IDs. You can define a custom device ID by setting it as a configuration option or by calling `setDeviceId`.

Reviewer: base64IDs or base64id? Can we align on this function?

Author: Fixed. Name should be "base64 ID" and the function is named "base64id()"

```js
amplitude.getInstance().setDeviceId('DEVICE_ID');
```

!!!note
    Amplitude doesn't recommend defining your own device IDs unless you have your own system for tracking user devices. Make sure the `deviceId` you set is unique to prevent conflicts with other devices in your Amplitude data.
    It's best practice to use something like a UUID.

    [See an example](https://github.com/amplitude/Amplitude-Javascript/blob/master/src/uuid.js) of how to generate UUIDs with JavaScript.

#### Get device ID

> You can retrieve the device ID that Amplitude uses with `Amplitude.getInstance().getDeviceId()` This method can return `null` if a `deviceId` hasn't been generated yet.

Reviewer: Question: I remembered we have a why to check if the deviceId has been set in android and ios. Does that exist in js SDK?

Author: Didn't find one in JS SDK, either in mobile sdks. Do you remember where it is? I can add it

Reviewer: NVM then. Thanks for checking.
